jane/doc/extensions/kinds.md

```diff
@@ -0,0 +1,335 @@
+details of the implementation of related features in the OCaml runtime. See the
+documentation for each mode to understand which types cross on its axis.
+
+Formally, these are called modal _bounds_ because the represent upper or lower
+bounds on the appropriate modal axes. For _future_ modal axes (like portability
+and linearity), the kind records an upper bound on the mode of values of this
+type. For example, `int` is `mod portable` because if you have an `int` that is
+`non-portable`, it's safe to treat it as `portable`.  For _past_ modal axes
+(like contention and uniqueness), the kind records a lower bound on
+_expectations_.  For example, `int` is `mod contended` because in a place where
+an `uncontended` value is expected, it's still safe to use a `contended` int.
+
+Why do past and future modal axes get different treatment in kinds? This is
+covered in the "Advanced Topics" section below, but isn't essential to
+understand for day-to-day use of the system.
+
+# Subkinding
+
+There is a partial order on kinds, which we'll write `k1 <= k2`. The
+relationship `k1 <= k2` holds when `k2` tells us less about a type than `k1`.
+Thus, it is always safe to use a type of kind `k1` where a type of kind `k2` is
+expected.  There is a maximum mode, written `any` (which, somewhat confusingly,
+is also the name of the maximum layout).
+
+As an example, `value mod portable <= value`. This means that if we know a type
+is represented as a normal OCaml value and mode crosses on the portability axis,
+it's safe to use the type where we just need a regular OCaml type but don't care
+how it interacts with portability. This relation forms a partial order because
+some kinds are unrelated, like `float64 mod contended` and `bits64 mod
+contended`, or `value mod portable` and `value mod aliased`.
+
+If you want to get nerdy about it, each individual piece of a kind (the layout
+and each possible axis of bounds) is a join semilattice, and the order we're
+talking about here is the one corresponding to the product join semilattice that
+combines all these components. The order for each component is described in the
+documentation for that component.
+
+# Kinds in the syntax
+
+You can write kind annotations in several places. In type declarations, they can appear
+both on the type being declared itself, and on the type parameters.  E.g.,:
+```ocaml
+type ('a : <kind1>) t : <kind2>
+```
+Kind annotations can also appear anywhere a type variable or underscore is legal
+in a type expression, as in:
+```ocaml
+let foo (x : ('a : value mod portable)) (y : (_ : bits32 & float64)) = ...
+```
+And on locally abstract types:
+```ocaml
+let foo (type (a : value mod portable) (b : bits32 & float64)) (x : a) (y : b)
+  = ...
+```
+
+It's legal to omit the `mod ...` portion of the kind when you only want to
+specify the layout (as we did in a couple of these examples). The bounds of
+unspecified axes are the maximum on those axes.
```

Comment:
This is subtle. When considered as a mode, `aliased` is the maximum on the uniqueness axis. But when considered as a modal bound, `unique` is the maximum: it mode-crosses the least. This means that `value mod aliased <= value mod unique` (because the latter is identically `value`), even though `unique <= aliased`.

This is sad and confusing, yet true of all past axes.

Reply:
I've added a paragraph explaining what "maximum" means here - let me know if you think that covers it.